firmware/controllers/engine_cycle/high_pressure_fuel_pump.cpp

```diff
@@ -13,70 +13,172 @@
 #include "high_pressure_fuel_pump.h"
 #include "spark_logic.h"
 
-#if EFI_HPFP
-
-#define LOBE_COUNT 3
+//#if EFI_HPFP
+#if 1
 
 class HpfpActor {
 public:
-	angle_t extra;
-	int phaseIndex;
+    uint8_t m_lobe_index;
+    float m_I_sum;
+    float m_pressureTarget;
+
+    AngleBasedEvent m_open;
+    AngleBasedEvent m_close;
 
-	AngleBasedEvent open;
-	AngleBasedEvent close;
 	DECLARE_ENGINE_PTR;
+
+    void start();
+
+private:
+    void scheduleNextCycle();
+    angle_t findNextLobe();
+    float calcFuelPercent(int rpm);
+    float calcPI(int rpm, float calc_fuel_percent);
+
+    static void hpfpPinTurnOn(HpfpActor *) {
+        enginePins.hpfpValve.setHigh();
+    }
+
+    static void hpfpPinTurnOff(HpfpActor *current) {
+        RegisteredNamedOutputPin *output = &enginePins.hpfpValve;
+        output->setLow();
+
+        current->scheduleNextCycle();
+    }
 };
 
-static HpfpActor actors[LOBE_COUNT];
+static HpfpActor actor; // TODO: Instantiate one per bank? Need more sensors....
```

Comment:
let's start with single pump control - 99% of the engines we care about for now are 4cyl, single pump, but on a variable cam